iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1097 @@
-            // after it had subscribed to the message topic.
-            _retainMessagesAcrossSessions = !settings.CleanSession;
+            List<X509Certificate> certs = _connectionCredentials.Certificate == null
+                ? new List<X509Certificate>(0)
+                : new List<X509Certificate> { _connectionCredentials.Certificate };
 
-            _webProxy = settings.Proxy;
+            tlsParameters.Certificates = certs;
 
-            if (channelFactory != null)
+            if (_mqttTransportSettings?.RemoteCertificateValidationCallback != null)
             {
-                _channelFactory = channelFactory;
+                tlsParameters.CertificateValidationHandler = certificateValidationHandler;
             }
-            else
+
+            tlsParameters.UseTls = true;
+            tlsParameters.SslProtocol = _mqttTransportSettings.SslProtocols;
+            _mqttClientOptionsBuilder.WithTls(tlsParameters);
+
+            _mqttClientOptionsBuilder.WithProtocolVersion(MQTTnet.Formatter.MqttProtocolVersion.V311); // 3.1.1
+
+            _mqttClientOptionsBuilder.WithCleanSession(_mqttTransportSettings.CleanSession);
+
+            _mqttClientOptionsBuilder.WithKeepAlivePeriod(_mqttTransportSettings.KeepAlive);
+
+            if (_mqttTransportSettings.HasWill && _mqttTransportSettings.WillMessage != null)
             {
-                _channelFactory = settings.Protocol switch
+                _mqttClientOptionsBuilder.WithWillTopic(_deviceToCloudMessagesTopic);
+                _mqttClientOptionsBuilder.WithWillPayload(_mqttTransportSettings.WillMessage.Payload);
+
+                if (_mqttTransportSettings.WillMessage.QualityOfService == QualityOfService.AtMostOnce)
                 {
-                    IotHubClientTransportProtocol.Tcp => CreateChannelFactory(connectionCredentials, _additionalClientInformation, settings),
-                    IotHubClientTransportProtocol.WebSocket => CreateWebSocketChannelFactory(connectionCredentials, _additionalClientInformation, settings),
-                    _ => throw new InvalidOperationException($"Unsupported transport setting: {settings.Protocol}"),
-                };
+                    _mqttClientOptionsBuilder.WithWillQualityOfServiceLevel(MqttQualityOfServiceLevel.AtMostOnce);
+                }
+                else if (_mqttTransportSettings.WillMessage.QualityOfService == QualityOfService.AtLeastOnce)
+                {
+                    _mqttClientOptionsBuilder.WithWillQualityOfServiceLevel(MqttQualityOfServiceLevel.AtLeastOnce);
+                }
             }
 
-            _closeRetryPolicy = new RetryPolicy(new TransientErrorIgnoreStrategy(), 5, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
+            _mqttClient.ApplicationMessageReceivedAsync += HandleReceivedMessage;
+            _mqttClient.DisconnectedAsync += HandleDisconnection;
```

Comment:
`+=` for events can cause memory leaks if they are never removed. Are you ensuring they are removed when necessary?

I haven't checked the code yet, but at least I need a reminder for myself to check.